tests/function_registry.cpp

```diff
@@ -0,0 +1,789 @@
+HINT: Candidate functions:
+  unique_int32(int32)
+)");
+  test("unique_int32", { CreateInt32(), CreateInt32() }, R"(
+ERROR: Cannot call function 'unique_int32' using argument types: (int32, int32)
+HINT: Candidate functions:
+  unique_int32(int32)
+)");
+}
+
+TEST_F(TestFunctionRegistryPopulated, unique_exact_args)
+{
+  test("unique_no_args", {}, unique_no_args_);
+}
+
+TEST_F(TestFunctionRegistryPopulated, unique_int8)
+{
+  test("unique_int8", { CreateInt8() }, unique_int8_);
+
+  test("unique_int8", { CreateInt16() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (int16)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateInt32() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (int32)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateInt64() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (int64)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateUInt8() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (uint8)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateUInt16() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (uint16)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateUInt32() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (uint32)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+
+  test("unique_int8", { CreateUInt64() }, R"(
+ERROR: Cannot call function 'unique_int8' using argument types: (uint64)
+HINT: Candidate functions:
+  unique_int8(int8)
+)");
+}
+
+TEST_F(TestFunctionRegistryPopulated, unique_int16)
+{
+  test("unique_int16", { CreateInt8() }, unique_int16_);
+  test("unique_int16", { CreateInt16() }, unique_int16_);
+
+  test("unique_int16", { CreateUInt8() }, unique_int16_);
+
+  test("unique_int16", { CreateInt32() }, R"(
+ERROR: Cannot call function 'unique_int16' using argument types: (int32)
+HINT: Candidate functions:
+  unique_int16(int16)
+)");
+
+  test("unique_int16", { CreateInt64() }, R"(
+ERROR: Cannot call function 'unique_int16' using argument types: (int64)
+HINT: Candidate functions:
+  unique_int16(int16)
+)");
+
+  test("unique_int16", { CreateUInt16() }, R"(
+ERROR: Cannot call function 'unique_int16' using argument types: (uint16)
+HINT: Candidate functions:
+  unique_int16(int16)
+)");
+
+  test("unique_int16", { CreateUInt32() }, R"(
+ERROR: Cannot call function 'unique_int16' using argument types: (uint32)
+HINT: Candidate functions:
+  unique_int16(int16)
+)");
+
+  test("unique_int16", { CreateUInt64() }, R"(
+ERROR: Cannot call function 'unique_int16' using argument types: (uint64)
+HINT: Candidate functions:
+  unique_int16(int16)
+)");
+}
+
+TEST_F(TestFunctionRegistryPopulated, unique_int32)
```

Comment:
I dunno - feels like a lot of boiler plate to test every iteration of `fits_into`.